src/backtrace/mod.rs

```diff
@@ -101,6 +101,14 @@ impl Frame {
     pub fn symbol_address(&self) -> *mut c_void {
         self.inner.symbol_address()
     }
+
+    /// Returns the register value at this frame.
+    ///
+    /// The general purpose register value may not be available. The register
+    /// index differs between platforms.
+    pub fn gr(&self, index: u32) -> Option<usize> {
```

Comment:
Could this perhaps be renamed to `general_register`? (or something similar)

The parameter/return value here is also pretty interesting since it's highly platform-specific. What about doing something like this:

```rust
// src/lib.rs
#[cfg(target_arch = "x86_64")]
pub mod x86_64 {
    pub struct Rax;
    // ...
    impl Register for Rax { /* ... */ }
}
// ...

impl Frame {
    pub fn general_register<T: Register>(&self) -> T::Value { /* ... */ }
}

pub trait Register {
    type Value;
    #[doc(hidden)] // internal details
    fn get(frame: &Frame) -> Self::Value;
}
```

That way we can extend this to all kinds of register (simd/32-bit/64-bit/etc), highlight how it's platform-specific (module names), and avoid the need for figuring fiddling with indices which are sourced from elsewhere

Reply:
It would be better to not make this generic and just use a `#[non_exhaustive]` enum for the registers.

Reply:
@Amanieu I don't understand what you mean, can you elaborate?

Reply:
Basically something like this:
```rust
#[cfg(target_arch = "x86_64")]
pub mod x86_64 {
    #[non_exhaustive]
    pub enum GeneralReg {
        Rax,
        ...
    }
    
    impl RegisterClass for GeneralReg { ... }
}

impl Frame {
    pub fn register<T: RegisterClass>(&self, reg: T) -> T::Value { /* ... */ }
}

pub trait RegisterClass {
    type Value;
    #[doc(hidden)] // internal details
    fn get(&self, frame: &Frame) -> Self::Value;
}
```

Reply:
That seems largely the same? Is there a reason to prefer one over the other?

Reply:
Using an `enum` allows for useful things like iterating over all available registers and printing them all. Doing this with only generics is not as ergonomic

```rust
use x86_64::GeneralReg::*;
for r in [Rax, Rbx, Rcx, ...] {
}
```

Reply:
Seems reasonable to me!

Could even add a cross-platform:

```rust
impl GeneralReg {
    pub fn all() -> impl Iterator<Item = GeneralReg> {
        // .. 
    }
}
```
or some sort of cross-platform iterator for "iterate over all the regs"